### PR TITLE
Feature/theming: Phase 1 — Bootstrap 2→3 (npm pipeline, Bootswatch)

### LINKED ISSUE #5807: Upgrade Bootstrap from 2.3.2 to 3.x
### Problem Statement

Currently using outdated Bootstrap 2.3.2. Need to move to 3.x as first step of sequential upgrade.

### Proposed Solution

Upgrade Bootstrap from 2.3.2 to 3.x. Refactor templates and update code to handle Bootstrap 3's breaking changes.

### Alternatives Considered

Upgrade directly to a later version, but iterative upgrades are likely to be less error-prone.

### Additional Context

This is a subissue of #3809. Bootstrap upgrade will be done incrementally.

### PR DESCRIPTION
**Pull request by Oval17**
## Description

Migrates the ESP theming system from **Bootstrap 2.3.2** to **Bootstrap 3.3.7**, replacing the 42 committed BS2 LESS source files with npm-managed dependencies and updating all six bundled themes plus shared templates to BS3 markup conventions.

This is Phase 1 of a 3-phase Bootstrap upgrade:
- **Phase 1 (this PR):** BS2 → BS3 (LESS pipeline, npm dependencies)
- **Phase 2:** BS3 → BS4 (LESS → SCSS migration, dart-sass)
- **Phase 3:** BS4 → BS5 (CSS custom properties, Bootstrap Icons, drop jQuery)

---

### Infrastructure

- **`package.json`** — replaced BS2 package metadata with `bootstrap: 3.3.7` + `bootswatch: 3.4.1` as npm dependencies
- **`ThemeController`** — updated LESS compile chain to resolve Bootstrap 3 from `node_modules/`; added `get_bootswatch_themes()`, updated `get_less_names()` + `compile_css()` for optional Bootswatch skin injection; fixed `get_variable_defaults()` to prepend `variables_custom.less` so theme variable files can reference compat variables
- **`variables_custom.less`** — added `@icon-font-path` (absolute path so glyphicons resolve from `/media/styles/`), `@link-color`/`@link-hover-color` bridges to BS3 kebab variables, `@navbarInverseLinkColorHover`, and BS2 compat shims (`@navbarInverseBackground` etc.)
- **`deploy/ci/before_script`** — `npm ci` runs before tests
- **`update_deps.sh`** — `npm ci` step added for production deploys
- **`docker-entrypoint.sh`** — auto-installs npm theme deps on container start if `node_modules` is absent (fixes fresh-clone setup)

### Template Migration — BS2 patterns removed

**Theme templates** (`esp/esp/themes/theme_data/*/`):
- `.row-fluid` → `.row`, `.spanN` → `.col-md-N`, `.offset-N` → `.col-md-offset-N`
- `.navbar-inner` removed, `<div class="navbar">` → `<nav class="navbar">`
- `.nav-collapse` → `.navbar-collapse`, `.btn-navbar` → `.navbar-toggle`
- `.nav-list` → `.nav.nav-stacked`
- `.accordion-heading/.accordion-body` → `.panel-heading/.panel-collapse`
- `.hero-unit` → `.jumbotron`
- `.icon-*` sprite classes → `.glyphicon.glyphicon-*`
- `.btn-inverse` → `.btn-default`
- `pull-left`/`pull-right` in navbars → `navbar-left`/`navbar-right`/`dropdown-menu-right`

**Shared templates** (`esp/templates/`):
- Migrated 17 files: `main.html`, `index.html`, `login.html`, `profile.html`, program module templates, theme editor templates — all `.row-fluid`/`.spanN`/`.offsetN` replaced

**Themes ported:** barebones, bigpicture, circles, droplets, floaty, fruitsalad (all 6)

### Bootswatch Integration

Backend support for optional Bootswatch 3 skins added to `ThemeController`. All 16 Bootswatch 3 themes (cerulean, cosmo, cyborg, darkly, flatly, journal, lumen, paper, readable, sandstone, simplex, slate, spacelab, superhero, united, yeti) are available via `compile_css(..., bootswatch_theme='cerulean')`. UI selector is deferred to Phase 2.

---

## Related Issue

Closes #5807 

## Type of Change

- [x] New feature
- [x] Refactor / cleanup
- [x] CI/CD or build change

## Testing

- [x] Existing tests pass (16/16 theme tests green; full CI passing)
- [x] New tests added — `Bootstrap3MigrationTest` (7 tests covering BS3 compile markers, Bootswatch discovery, LESS import order, `get_variable_defaults()` roundtrip for all themes)
- [x] Manually verified — all 6 themes compile and render correctly; navbar, sidebar, dropdowns, glyphicons, footer all verified at desktop and mobile widths

## Checklist

- [x] Self-reviewed the code
- [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by willgearty** [comment]
Thanks for putting together this first PR! I'm just going to run a quick Copilot review to catch any of the obvious issues. Also looks like there are a couple of failing tests.

**Comment by Oval17** [comment]
Thanks @willgearty for creating issues 😅, I was going to ask for same.
Seems there are few around 3 small comments plus test cases failing , I will resolve them and I will ping once I have completed all testing from my side.
Thanks
cc : @kkbrum

File: esp/public/media/theme_editor/package.json [hunk 1]
```diff
@@ -1,28 +1,10 @@
...
-    , "serve-static": "^1.15.0"
-    , "hogan.js": "^3.0.2"
+  "name": "esp-theme-editor",
+  "description": "Bootstrap and theme editor dependencies for ESP-Website",
+  "version": "3.0.0",
+  "private": true,
+  "dependencies": {
+    "bootstrap": "3.3.7",
+    "bootswatch": "3.4.1"
   }
```
[the remark this answers is not in the document]
Comment (Oval17) [reply esp/public/media/theme_editor/package.json:9]: Fixed

File: esp/esp/themes/theme_data/droplets/templates/users/loginbox_content.html [hunk 2]
```diff
@@ -16,7 +16,7 @@
   </li>
   <li>
     <form class="navbar-form form-inline" action="/myesp/signout">
-      <a href="/myesp/signout"><button class="btn btn-inverse" style="margin-right:0px;">Logout</button></a>
+      <a href="/myesp/signout"><button class="btn btn-default" style="margin-right:0px;">Logout</button></a>
     </form>
```
[the remark this answers is not in the document]
Comment (Oval17) [reply esp/esp/themes/theme_data/droplets/templates/users/loginbox_content.html]: This is not related to this pr , this is preexisiting and not required to be changed

File: deploy/ci/before_script [hunk 3]
```diff
@@ -4,6 +4,9 @@ set -euf -o pipefail
 if [ "$CI_JOB" = "test" ]; then
     cp esp/esp/local_settings.py.ci esp/esp/local_settings.py
 
+    # Install Bootstrap 3 and other npm dependencies for theme compilation
+    (cd esp/public/media/theme_editor && npm install)
```
[the remark this answers is not in the document]
Comment (Oval17) [reply deploy/ci/before_script]: Fixed

File: esp/public/media/theme_editor/less/variables_custom.less [hunk 4]
```diff
@@ -18,9 +28,8 @@
...
 @navbarInverseBackground:                #111111;
 @navbarInverseBackgroundHighlight:       #222222;
-@navbarInverseText:                      @grayLight;
-@navbarInverseLinkColor:                 @grayLight;
-@navbarInverseLinkColorHover:            @white;
+@navbarInverseText:                      #999;
+@navbarInverseLinkColor:                 #999;
 @navbarInverseBrandColor:                @navbarInverseLinkColor;
 @navbarBrandFontSize:       24px;
 @sidebarText:               #999;
```
[the remark this answers is not in the document]
Comment (Oval17) [reply esp/public/media/theme_editor/less/variables_custom.less:43]: Fixed

**Comment by Oval17** [comment]
Have fixed TCs , have addressed review comments , and have throughly tested and reviewed all the phase - 1 changes, 
Ready for review and inputs
meanwhile I will start working on phase-2 changes exploration and all the requirements...
Thanks

**Review by kkbrum — COMMENTED** [review]
Thanks for this! When I try to select a new theme or install updates for the theme, I get "FileNotFoundError: [Errno 2] No such file or directory: '/app/esp/public/media/theme_editor/node_modules/bootstrap/less/bootstrap.less'". I did docker compose build --no-cache. Is there anything else I need to do to be able to test this?

**Comment by Oval17** [comment]
> Thanks for this! When I try to select a new theme or install updates for the theme, I get "FileNotFoundError: [Errno 2] No such file or directory: '/app/esp/public/media/theme_editor/node_modules/bootstrap/less/bootstrap.less'". I did docker compose build --no-cache. Is there anything else I need to do to be able to test this?

When we do Fresh build and if theres no node modules mounted in machine for docker to pick this error is coming have verified this.
I have baked node in docker itself , this edge case wont pop again...
can you see now once....

**Comment by willgearty** [comment]
We also need to make sure this gets installed when we update our production sites. The directives probably need to go in here: https://github.com/learning-unlimited/ESP-Website/blob/main/esp/packages_base_manual_install.sh (called by deploy/ci/install)
or here: https://github.com/learning-unlimited/ESP-Website/blob/main/esp/update_deps.sh (only called for production)

**Review by kkbrum — CHANGES_REQUESTED** [review]
Thanks for the changes; it looks like it's taken effect in my docker container now. 

So far, I've only looked at droplets, but I'm seeing a lot of issues. Here's a non-exhaustive list:

- the various alert icons aren't showing:
Image

- the bottom of the pages are cut off:
Image

- the icons on the main program management pages aren't showing:
Image

- the admin toolbar is funny:
Image

- the navbar has dropdown arrows as their own row instead of next to the headings:
Image

**Comment by Oval17** [comment]
This should be fixed now, related to glphyicons not referenced properly.

**Comment by Oval17** [comment]
I am realising 
3 clean migration makes sense , instead of porting different ui components 

bs2-to-bs3 — current PR (almost done)
bs3-to-bs4 — LESS → Sass migration, Bootstrap 4 classes
bs4-to-bs5 — Bootstrap 5, drop jQuery dependency, CSS variables
Then a separate extended-theming for dark mode, Bootswatch UI selector, custom theme builder improvements etc.

This is very close to what we are doing it will only ease more things 
just a bit more effort will be needed from my side but will sort out most of ui related things.
We should go ahead with this.

**Comment by Oval17** [comment]
Hi @kkbrum 
I have tested this again.
There could few more UI issues, but I have tried to cover most of them.
I was actually testing with more changes that were not in pr like referenced glyphicons and row-col etc.
Thats why for the most part everything looked correct to me 
It was due to phase - 2 mixup , 
But I have now simplified plan as I have stated above.



Now things should be pretty straightforward.
Still a bit of thorough UI UX testing will be required, will fix if any more issues are there
Thanks.


Screenshot 2026-05-29 at 2 52 54 PM
Screenshot 2026-05-29 at 2 54 00 PM

File: esp/esp/themes/controllers.py [hunk 5]
```diff
@@ -282,18 +339,13 @@
...
-            less_file = open(filename)
-            logger.debug('Including LESS source %s', filename)
-            less_data += '\n' + less_file.read()
-            less_file.close()
-
-        #   Make icon image path load from the CDN by default
-        if 'iconSpritePath' not in variable_data:
-            variable_data['iconSpritePath'] = f'"{settings.CDN_ADDRESS}/bootstrap/img/glyphicons-halflings.png"'
+        for filename in self.get_less_names(theme_name, bootswatch_theme=bootswatch_theme):
+            with open(filename) as less_file:
```
[the remark this answers is not in the document]
Comment (Oval17) [reply esp/esp/themes/controllers.py]: Resolved

File: esp/esp/db/forms.py [hunk 6]
```diff
@@ -112,7 +112,7 @@ def render(self, name, value, attrs=None, renderer=None):
   <img src="/static/admin/img/search.svg" border="0" width="16" height="16" alt="Lookup" /></a>
    &nbsp;<strong>%(old_init_val)s</strong>
 </div>
-<input type="text" id="id_%(fn)s" name="%(fn)s_raw" value="%(data)s" class="span6" />
+<input type="text" id="id_%(fn)s" name="%(fn)s_raw" value="%(data)s" class="col-md-6" />
```
Comment (Oval17) [inline esp/esp/db/forms.py:115]: New format added here

File: esp/esp/program/modules/forms/rapidcheckin.py [hunk 7]
```diff
@@ -27,7 +27,7 @@ def render(self, name, value, attrs=None, renderer=None):
             except (ESPUser.DoesNotExist, ValueError, TypeError):
                 pass
         html = (
-            '<input type="text" id="id_%s" name="%s_raw" value="%s" class="span6" />'
+            '<input type="text" id="id_%s" name="%s_raw" value="%s" class="col-md-6" />'
```
Comment (Oval17) [inline esp/esp/program/modules/forms/rapidcheckin.py:30]: New format added here

File: esp/esp/themes/theme_data/barebones/scripts/sidebar.js [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
-        $j("#sidebar").addClass("span3");
+        $j("#sidebar").addClass("col-md-3");
     } else {
         $j("#sidebar").addClass("hidden");
-        $j("#sidebar").removeClass("span3");
-        $j("#main.resizable").removeClass("span9");
-        $j("#main.resizable").addClass("span12");
+        $j("#sidebar").removeClass("col-md-3");
+        $j("#main.resizable").removeClass("col-md-9");
+        $j("#main.resizable").addClass("col-md-12");
```
Comment (Oval17) [inline esp/esp/themes/theme_data/barebones/scripts/sidebar.js:11]: New format added here

File: esp/esp/themes/theme_data/barebones/templates/index.html [hunk 9]
```diff
@@ -4,10 +4,10 @@
...
 {% block content %}
 
-{% block heading %}  
-<div class="row-fluid">
-  <div class="span12">
-    <div class="hero-unit">
+{% block heading %}
+<div class="row">
+  <div class="col-md-12">
+    <div class="jumbotron">
```
Comment (Oval17) [inline esp/esp/themes/theme_data/barebones/templates/index.html:10]: New format added here

File: esp/esp/themes/theme_data/barebones/templates/index.html [hunk 10]
```diff
@@ -17,18 +17,18 @@ <h1>Welcome to {{ settings.ORGANIZATION_SHORT_NAME }}</h1>
...
+  <div class="col-md-4">
     <a href="/teach/index.html"><h2>Teach</h2></a>
-        <p>Edit this page to serve as a starting point for your teachers.</p>
+    <p>Edit this page to serve as a starting point for your teachers.</p>
   </div>
-  <div class="span4">
+  <div class="col-md-4">
     <a href="/volunteer/index.html"><h2>Volunteer</h2></a>
-        <p>Edit this page to serve as a starting point for your volunteers.</p>
+    <p>Edit this page to serve as a starting point for your volunteers.</p>
```
Comment (Oval17) [inline esp/esp/themes/theme_data/barebones/templates/index.html:31]: New format added here

File: esp/esp/themes/theme_data/barebones/templates/main.html [hunk 11]
```diff
@@ -35,7 +35,7 @@
   <div class="row">
     {% block sidebar %}
     <div class="hidden" id="sidebar">
-      <ul class="nav nav-list sidebar">
+      <ul class="nav nav-stacked sidebar">
```
Comment (Oval17) [inline esp/esp/themes/theme_data/barebones/templates/main.html:38]: BS3 compatible Nav

File: esp/esp/themes/theme_data/barebones/templates/main.html [hunk 12]
```diff
@@ -44,7 +44,7 @@
     </div>
     {% endblock %}
 
-    <main id="main" class="span12 resizable" tabindex="-1">
+    <main id="main" class="col-md-12 resizable" tabindex="-1">
```
Comment (Oval17) [inline esp/esp/themes/theme_data/barebones/templates/main.html:47]: New format added here

File: esp/esp/themes/theme_data/bigpicture/less/main.less [hunk 13]
```diff
@@ -126,22 +126,20 @@ div.nav-collapse.collapse {
         font-size: 17px;
     }
     .active {
-        [class^="icon-"], [class*=" icon-"] {
-            background-image: url("/media/theme_editor/img/glyphicons-halflings-white.png");
+        .glyphicon {
```
Comment (Oval17) [inline esp/esp/themes/theme_data/bigpicture/less/main.less:129]: .glyphicon resolved

File: esp/esp/themes/theme_data/bigpicture/templates/users/loginbox_content.html [hunk 14]
```diff
@@ -34,7 +34,7 @@
         <button class="btn dropdown-toggle" data-toggle="dropdown" style="margin-left:0px;">
           &#9662;
         </button>
-        <ul class="dropdown-menu pull-right">
+        <ul class="dropdown-menu dropdown-menu-right">
```
Comment (Oval17) [inline esp/esp/themes/theme_data/bigpicture/templates/users/loginbox_content.html:37]: same as above

File: esp/esp/themes/theme_data/droplets/templates/navbar/admin.html [hunk 15]
```diff
@@ -31,7 +31,7 @@
     </ul>
   </li>
 </ul>
-<ul class="nav pull-left unmorph hidden">
+<ul class="nav navbar-nav unmorph hidden">
```
Comment (Oval17) [inline esp/esp/themes/theme_data/droplets/templates/navbar/admin.html:34]: same as above

File: esp/templates/users/userview.html [hunk 16]
```diff
@@ -211,8 +211,8 @@ <h2>
     <span class="edit" style="display:none;">
         <input type="text" id="edit_first_name" value="{{ user.first_name }}" placeholder="First Name" />
         <input type="text" id="edit_last_name" value="{{ user.last_name }}" placeholder="Last Name" />
-        <button class="btn btn-small btn-primary" onclick="submitEditUser('name');">Save</button>
-        <button class="btn btn-small btn-default" onclick="toggleEditUser('name', false); return false;">Cancel</button>
+        <button class="btn btn-sm btn-primary" onclick="submitEditUser('name');">Save</button>
+        <button class="btn btn-sm btn-default" onclick="toggleEditUser('name', false); return false;">Cancel</button>
```
Comment (Oval17) [inline esp/templates/users/userview.html:215]: new format added here

**Review by Oval17 — COMMENTED** [review]
Have added npm paths for production as well
Most of the new format looks good 
There is around 15k lines being removed which will be used dynamically with bs3(though it wont stay with us for long😅)

**Review by kkbrum — CHANGES_REQUESTED** [review]
This is looking a lot better. Testing some other themes now--

- It looks like the button customization is broken. It seems like inverse buttons no longer exist in bootstrap 3, so they need to get removed from the theme editor. However, even the primary and default button customization doesn't seem to work, at least in fruitsalad:

Image

- I know we don't care much about barebones, but the header is overlapping some text right now. Could we fix that?

Image

**Comment by Oval17** [comment]
> This is looking a lot better. Testing some other themes now--
> 
> * It looks like the button customization is broken. It seems like inverse buttons no longer exist in bootstrap 3, so they need to get removed from the theme editor. However, even the primary and default button customization doesn't seem to work, at least in fruitsalad:
> 
> Image
> * I know we don't care much about barebones, but the header is overlapping some text right now. Could we fix that?
> 
> Image

These should be fixed now...
@kkbrum

**Review by kkbrum — APPROVED** [review]
Great! I approve this for merging once the merge conflicts are addressed. @Oval17 could you take care of those?

**Comment by Oval17** [comment]
> Great! I approve this for merging once the merge conflicts are addressed. @Oval17 could you take care of those?

Yeah these are due to earlier changes I did regarding imports resolving this.

**Comment by Oval17** [comment]
We can go ahead with the merge, i guess your approval is required as well @willgearty 
Thanks

**Comment by willgearty** [comment]
🙌
